Toolchain/ComputeDependenciesHash.sh

```diff
@@ -23,89 +23,106 @@ function finish {
-        -e 's^#include <(Lib[A-Za-z]+/.*)>^Libraries/\1^' \
-        -e 's^#include <((bits|netinet|sys|arpa|net)/.*)>^Libraries/LibC/\1^' \
-        -e 's^#include <fd_set.h>^Libraries/LibC/fd_set.h^' \
-        -e 's^#include <([a-z]{3,10}(_numbers)?\.h)>^Libraries/LibC/\1^' \
-        -e 's^#include <([A-Z][a-z]+Server/.*)>^Services/\1^' \
-        -e 's^#include <(.*)>^UNRESOLVED_I/\1^' \
-        -e 's^#include "(.*)"^UNRESOLVED_L/\1^' > "${DEPLIST_FILE}"
-# Some #include's cannot be resolved, like <chrono>. However, these are only
-# a problem if they turn up as a transitive dependency of libc and libm.
-# We will check for that when the time comes.
+# First, capture the caller's input.
+echo "$0: Configuration:" >&2
+cat /dev/stdin | tee /dev/stderr > "${DEPLIST_FILE}"
+# "$@" is the md5sum invocation.
+"$@" Toolchain/ComputeDependenciesHash.sh | tee /dev/stderr > "${DEPLIST_FILE}"
 
-# The initial guess is pessimistic: *all* of libc and libm.
-FILE_LIST=$(find Libraries/LibC/ Libraries/LibM/ \( -name '*.cpp' -o -name '*.c' -o -name '*.h' \) ! -name 'Test*')
-echo "$0: Exploring dependencies of libstdc++" >&2
-FILE_LIST_COMPLETE="n"
-# In each iteration, we extend FILE_LIST by the dependencies not listed yet in
-# FILE_LIST.  Note that the results are always semantically the same,
-# but the order depends on the initial `find` runs.
-for _ in $(seq 10) ; do
-    FILE_REGEX=$(echo "${FILE_LIST}" | sed -zr -e 's,\n$,,' -e 's,\.,\\.,g' -e 's,\n,|,g')
-    FURTHER_FILE_LIST=$(grep -P "^(${FILE_REGEX}):" "${DEPLIST_FILE}" | grep -Pv ":(${FILE_REGEX})\$" | sed -re 's,^.*:(.*)$,\1,' | sort -u)
-    if [ -n "${FURTHER_FILE_LIST}" ] ; then
-        # FILE_LIST should grow to a maximum of "number of all .cpp and .c and .h files",
-        # i.e. roughly 700 lines.  This should be managable, even as the project grows.
-        FILE_LIST="${FILE_LIST}
-${FURTHER_FILE_LIST}"
+# libstdc++ depends on the *headers* of libc, so we pessimistically assume it depends
+# on *all* of the files that `make install-target-libstdc++-v3` looks at.
+# This list of files was gathered by running:
+#     strace -ff -e trace=file "make install-target-libstdc++-v3" 2>&1 >/dev/null | perl -ne 's/^[^"]+"(([^\\"]|\\[\\"nt])*)".*/$1/ && print' | sort -u | grep -P 'serenity/Build/Root/usr/include/.*\.h$'
+# This might change in the future, but I *extremely* doubt it.
+#
+# Note that this list contains files that don't exist in "our" LibC.
+# This is fine! We might need to recompile the toolchain if those headers define stuff that libstdc++ recognizes.
+# Note that this list don't contain all files of "our" LibC.
+# This is fine! libstdc++ most definitely doesn't depend on those symbols then.
+# Note that these files include files not on the list.
+# This is ... probably fine? The files are currently bits/endian.h, sched.h, stdarg.h, stdbool.h, stddef.h.
+# They are sufficiently unlikely to change breakingly. And if they do, then one of the files below will change anyway.
```

Comment:
Wouldn't it be easier and more future-proof to just enumerate all `.h` files on our side?

Reply:
- Advantage: "Future-proof", in case of changes that change the very small part that libstdc++ cares about. If this happens, then one of the other files will change, so I don't see the point.
- Disadvantage: False positives all over the place, like before. The point is to prevent Travis from constantly rebuilding the Toolchain, isn't it?

Reply:
I mean `.h` files in our LibC. Wouldn't that be very narrow anyway, plus cover future GCC updates?
Or maybe I'm missing something obvious here.

Reply:
Future gcc updates are covered anyway, as `BuildIt.sh` (which contains the gcc version string) and the gcc version string are (among other things) hashed into the "Toochain Hash", together with the files we're currently discussion. So gcc is already handled at this point.

Regarding `.h` files in our LibC: That's what the above code does: hashes most of our `.h` files, so that the Toolchain will be rebuilt whenever any of the files change, no matter how. It's only "most" files because libstdc++ doesn't look at the other `.h` files, and `strace`ing the build and installation of libstdc++ proves this.

So … I think you're suggesting to do what the code already does? :)

Reply:
Well, I just see a bit static list of files and it makes me wonder if it could be dynamically generated, that's all :)

Reply:
It can! You just need to `strace` the build process of `libstdc++`, which is a little bit slower than "just" building `libstdc++` directly, and also requires a freshly built gcc. That's why I decided against doing it dynamically. :P

Reply:
What I don't understand is: why is the static list better than a simple `find Libraries/LibC -type f -name '*.h'`?
Also note that future changes to our LibC headers may cause the `libstdc++` configure script to reach different conclusions about feature availability and thus alter the dependency graph.

Reply:
Because libstdc++ by far doesn't care about all of our headers, and I expect the probability of this happening *and* none of the `.h` changing to be negligible.

But okay, I'll change it: It seems we have very different estimates for "cost of unnecessary rebuild" and "probability of unrecognized ABI change" very differently.